### PR TITLE
[Qt] - update to 5.15.2

### PR DESCRIPTION
**Pull request by ViralBShah**
Qt update. @barche 

The build script is looking solid!

### CODE REVIEW

**Comment by ViralBShah** [comment]
I guess we should wait for the Pkg issue to be resolved.

**Comment by barche** [comment]
OK, if it's this one we're waiting for, it is merged: https://github.com/JuliaLang/Pkg.jl/pull/2273

File: Q/Qt/build_tarballs.jl [hunk 1]
```diff
@@ -47,15 +47,15 @@ case "$target" in
             -skip qtwinextras -fontconfig
 #        ln -s /lib64/libc.so.6 /lib64/libc.so
```
Comment (giordano) [inline Q/Qt/build_tarballs.jl:48]:
```suggestion
```

Comment (ViralBShah) [reply Q/Qt/build_tarballs.jl:48]: I thought I would leave it in there should it come up in the future.

**Comment by ViralBShah** [comment]
> OK, if it's this one we're waiting for, it is merged: [JuliaLang/Pkg.jl#2273](https://github.com/JuliaLang/Pkg.jl/pull/2273)

Not sure if we need to wait until it makes its way into master. @giordano Is it ok to merge stuff?

File: Q/Qt/build_tarballs.jl (hunk 1 above)
Comment (giordano) [reply Q/Qt/build_tarballs.jl]: Having to do that feels mostly wrong to me :slightly_smiling_face:

**Comment by giordano** [comment]
I don't understand how the linked Pkg PR is related with this PR at all

**Comment by ViralBShah** [comment]
Not sure which PR - but there is the tree mismatch error on master - and I was wondering if we need to wait until it is resolved. I assume nothing has changed here in Yggdrasil. Maybe we are ok to keep merging here?

**Comment by giordano** [comment]
> but there is the tree mismatch error on master 

It's for `GR_jll` (I didn't see anyone complaining about it for `Qt_jll`), and it's just a Pkg bug, not related with what we do here: waiting for https://github.com/JuliaLang/Pkg.jl/pull/2273 to make its way in Julia master wouldn't change anything (also because here we don't use Julia master, which is totally broken for BinaryBuilder)

**Comment by giordano** [comment]
> Maybe we are ok to keep merging here?

In general yes, at this point I don't see reasons not to go on with this

**Comment by ViralBShah** [comment]
I'm still having the file permissions issues with this version of Qt when I delete my `.julia`. Perhaps the relevant PR is not yet merged?

**Comment by giordano** [comment]
Are you sure the directory being deleted is the one of version 5.15.2?

**Comment by ViralBShah** [comment]
I wiped my `.julia`, then installed `GR`, which pulled in the latest Qt release (5.15.2).

**Comment by giordano** [comment]
Seeing the full path of the offending files would help to investigate the issue

**Comment by giordano** [comment]
In #1969 you showed that files in `lib/QtMultimediaQuick.framework` had wrong permissions in the old tarball, but in the new one I get:
```
% tar tzvf Qt.v5.15.2.x86_64-apple-darwin.tar.gz|grep "QtMultimediaQuick.framework"
lrwxr-xr-x 0/0               0 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Headers -> Versions/Current/Headers
lrwxr-xr-x 0/0               0 1970-01-01 01:00 lib/QtMultimediaQuick.framework/QtMultimediaQuick -> Versions/Current/QtMultimediaQuick
lrwxr-xr-x 0/0               0 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Resources -> Versions/Current/Resources
-rw-r--r-- 0/0            4483 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Versions/5/Headers/5.15.2/QtMultimediaQuick/private/qdeclarativevideooutput_backend_p.h
-rw-r--r-- 0/0            7271 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Versions/5/Headers/5.15.2/QtMultimediaQuick/private/qdeclarativevideooutput_p.h
-rw-r--r-- 0/0            5267 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Versions/5/Headers/5.15.2/QtMultimediaQuick/private/qdeclarativevideooutput_render_p.h
-rw-r--r-- 0/0            3142 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Versions/5/Headers/5.15.2/QtMultimediaQuick/private/qdeclarativevideooutput_window_p.h
-rw-r--r-- 0/0            3990 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Versions/5/Headers/5.15.2/QtMultimediaQuick/private/qsgvideonode_p.h
-rw-r--r-- 0/0            3208 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Versions/5/Headers/5.15.2/QtMultimediaQuick/private/qsgvideonode_rgb_p.h
-rw-r--r-- 0/0            3225 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Versions/5/Headers/5.15.2/QtMultimediaQuick/private/qsgvideonode_texture_p.h
-rw-r--r-- 0/0            3256 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Versions/5/Headers/5.15.2/QtMultimediaQuick/private/qsgvideonode_yuv_p.h
-rw-r--r-- 0/0            2532 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Versions/5/Headers/5.15.2/QtMultimediaQuick/private/qtmultimediaquickdefs_p.h
-rw-r--r-- 0/0             175 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Versions/5/Headers/QtMultimediaQuick
-rw-r--r-- 0/0             268 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Versions/5/Headers/QtMultimediaQuickDepends
-rw-r--r-- 0/0              38 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Versions/5/Headers/QtMultimediaQuickVersion
-rw-r--r-- 0/0             253 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Versions/5/Headers/qtmultimediaquickversion.h
-rwxr-xr-x 0/0          183000 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Versions/5/QtMultimediaQuick
-rw-r--r-- 0/0             660 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Versions/5/Resources/Info.plist
-rw-r--r-- 0/0            1959 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Versions/5/Resources/QtMultimediaQuick.prl
lrwxr-xr-x 0/0               0 1970-01-01 01:00 lib/QtMultimediaQuick.framework/Versions/Current -> 5
```
If some files are still saved with read-only permissions, must be something else, but
```
tar tzvf Qt.v5.15.2.x86_64-apple-darwin.tar.gz|grep "^-r-"
```
doesn't find anything

**Comment by ViralBShah** [comment]
Here are some files I can't delete. Pretty much everything gets installed with these permissions.

```
/Users/viral/.julia/artifacts/fc14947a44c2b7772250cff86ab2df52046e63d3/lib/QtConcurrent.framework/Versions/5/Headers
➜  Headers ls -l
total 616
-r--r--r--  1 viral  staff    685 Dec  8 13:50 QtConcurrent
-r--r--r--  1 viral  staff    180 Dec  8 13:50 QtConcurrentDepends
-r--r--r--  1 viral  staff     32 Dec  8 13:50 QtConcurrentFilter
-r--r--r--  1 viral  staff     29 Dec  8 13:50 QtConcurrentMap
-r--r--r--  1 viral  staff     29 Dec  8 13:50 QtConcurrentRun
-r--r--r--  1 viral  staff     33 Dec  8 13:50 QtConcurrentVersion
-r--r--r--  1 viral  staff   2289 Dec  8 13:50 qtconcurrent_global.h
-r--r--r--  1 viral  staff   2465 Dec  8 13:50 qtconcurrentcompilertest.h
-r--r--r--  1 viral  staff   2385 Dec  8 13:50 qtconcurrentexception.h
-r--r--r--  1 viral  staff   9700 Dec  8 13:50 qtconcurrentfilter.h
-r--r--r--  1 viral  staff  11885 Dec  8 13:50 qtconcurrentfilterkernel.h
-r--r--r--  1 viral  staff  10001 Dec  8 13:50 qtconcurrentfunctionwrappers.h
-r--r--r--  1 viral  staff  10164 Dec  8 13:50 qtconcurrentiteratekernel.h
-r--r--r--  1 viral  staff  11626 Dec  8 13:50 qtconcurrentmap.h
-r--r--r--  1 viral  staff   9810 Dec  8 13:50 qtconcurrentmapkernel.h
-r--r--r--  1 viral  staff   5994 Dec  8 13:50 qtconcurrentmedian.h
-r--r--r--  1 viral  staff   7666 Dec  8 13:50 qtconcurrentreducekernel.h
-r--r--r--  1 viral  staff  65328 Dec  8 13:50 qtconcurrentrun.h
-r--r--r--  1 viral  staff   4296 Dec  8 13:50 qtconcurrentrunbase.h
-r--r--r--  1 viral  staff  97481 Dec  8 13:50 qtconcurrentstoredfunctioncall.h
-r--r--r--  1 viral  staff   8410 Dec  8 13:50 qtconcurrentthreadengine.h
-r--r--r--  1 viral  staff    228 Dec  8 13:50 qtconcurrentversion.h
```

**Comment by ViralBShah** [comment]
Perhaps a mac specific issue?

**Comment by giordano** [comment]
This is what's in the tarball:
```
% tar tzvf Qt.v5.15.2.x86_64-apple-darwin.tar.gz|grep "QtConcurrent.framework/Versions/5/Headers"
-rw-r--r-- 0/0             685 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/QtConcurrent
-rw-r--r-- 0/0             180 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/QtConcurrentDepends
-rw-r--r-- 0/0              32 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/QtConcurrentFilter
-rw-r--r-- 0/0              29 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/QtConcurrentMap
-rw-r--r-- 0/0              29 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/QtConcurrentRun
-rw-r--r-- 0/0              33 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/QtConcurrentVersion
-rw-r--r-- 0/0            2289 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/qtconcurrent_global.h
-rw-r--r-- 0/0            2465 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/qtconcurrentcompilertest.h
-rw-r--r-- 0/0            2385 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/qtconcurrentexception.h
-rw-r--r-- 0/0            9700 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/qtconcurrentfilter.h
-rw-r--r-- 0/0           11885 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/qtconcurrentfilterkernel.h
-rw-r--r-- 0/0           10001 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/qtconcurrentfunctionwrappers.h
-rw-r--r-- 0/0           10164 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/qtconcurrentiteratekernel.h
-rw-r--r-- 0/0           11626 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/qtconcurrentmap.h
-rw-r--r-- 0/0            9810 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/qtconcurrentmapkernel.h
-rw-r--r-- 0/0            5994 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/qtconcurrentmedian.h
-rw-r--r-- 0/0            7666 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/qtconcurrentreducekernel.h
-rw-r--r-- 0/0           65328 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/qtconcurrentrun.h
-rw-r--r-- 0/0            4296 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/qtconcurrentrunbase.h
-rw-r--r-- 0/0           97481 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/qtconcurrentstoredfunctioncall.h
-rw-r--r-- 0/0            8410 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/qtconcurrentthreadengine.h
-rw-r--r-- 0/0             228 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/qtconcurrentversion.h
```
:woman_shrugging:

**Comment by ViralBShah** [comment]
Confirmed - not an issue on Linux, only on macOS.

**Comment by giordano** [comment]
`BinaryBuilder` is responsible for the production of the tarball, then `Pkg` unpacks it (probably using `Tar.jl` under the hood if you're on julia master).  As I've shown above, the macOS tarball has now the correct file permissions, with nothing in read-only mode for the user.

**Comment by ViralBShah** [comment]
It's odd that the problem only happens for the Qt tarball and no others. 

@StefanKarpinski I suspect we should file this mac permissions issue somewhere, but not sure if it should be in Pkg.jl or Tar.jl.

**Comment by ViralBShah** [comment]
> This is what's in the tarball:
> 
> ```
> % tar tzvf Qt.v5.15.2.x86_64-apple-darwin.tar.gz|grep "QtConcurrent.framework/Versions/5/Headers"
> -rw-r--r-- 0/0             685 1970-01-01 01:00 lib/QtConcurrent.framework/Versions/5/Headers/QtConcurrent
```
Not that it matters, but why do we have a 1970 date?

**Comment by giordano** [comment]
> Not that it matters, but why do we have a 1970 date in the tarball? 

Gotta ask @StefanKarpinski
```
julia> mktempdir() do dir
           mkpath(joinpath(dir, "dir"))
           touch(joinpath(dir, "dir", "file"))
           run(`tar tvf $(Tar.create(dir))`)
       end;
-rw-r--r-- 0/0               0 1970-01-01 01:00 dir/file
```
Maybe recording the timestamp isn't that important and so we just use the Unix epoch?

**Comment by giordano** [comment]
The `README` explains the rationale for not recording timestamps in a few places, in particular https://github.com/JuliaIO/Tar.jl/blob/0ba9683b202f09fe13b18368d6f6f20eca500c92/README.md#time-stamps

**Comment by StefanKarpinski** [comment]
We want tarballs to reproducible and purely a function of their content, so you don't want to include the time at which it was created or other incidental junk from your local system like the name of the user that owns the file. Coaxing `tar` into making reproducible tarballs is actually quite difficult and was one of the reasons to create `Tar.jl`.

**Comment by StefanKarpinski** [comment]
There are two possible issues here:

1. There was a brief window of time where `Pkg` was using `Sys.isexecutable` on all platforms to detect whether a file was user-executable, which does not work reliably on macOS; that has since been fixed, so if you're on a current version of Julia, it should not happen.

2. If the tarball has two paths that differ only by case, then one of them will overwrite the other on a case-insensitive file system like macOS (or Windows, I believe), so the tree hash of the extracted content will not match the intended hash of the tarball. We should prevent registration of packages and artifacts that have this property.

**Comment by StefanKarpinski** [comment]
If you can provide me with a link to the problematic tarball, I can check if Tar.jl is extracting it correctly on macOS.

**Comment by giordano** [comment]
> If the tarball has two paths that differ only by case, then one of them will overwrite the other on a case-insensitive file system like macOS (or Windows, I believe)

I'm moderately sure we have guards against this in BinaryBuilder, at least as warnings in the audit

> so the tree hash of the extracted content will not match the intended hash of the tarball. 

I'm a bit confused by this comment: the problem observed by Viral is that some files get wrong permissions (read-only) when unpackaged, but the tarball looks ok.

> If you can provide me with a link to the problematic tarball, I can check if Tar.jl is extracting it correctly on macOS.

https://github.com/JuliaBinaryWrappers/Qt_jll.jl/releases/download/Qt-v5.15.2%2B0/Qt.v5.15.2.x86_64-apple-darwin.tar.gz (but see above)

**Comment by StefanKarpinski** [comment]
> I'm a bit confused by this comment: the problem observed by Viral is that some files get wrong permissions (read-only) when the artifact is unpacked, but the tarball looks ok.

I hadn't read the earlier part of the thread when I posted that. There were two issues reported at about the same time, this is what they were.

**Comment by barche** [comment]
I tried manually extracting using Tar.jl, and the resulting file tree has no permission issues. But the Pkg-extracted files do have the wrong permissions, both in Julia 1.5 and in a freshly downloaded nightly.

**Comment by StefanKarpinski** [comment]
Ah, looks like https://github.com/JuliaLang/julia/pull/38755 wasn't merged yet, which may be the reason.

**Comment by giordano** [comment]
But that would help only in Julia v1.6.  `Pkg.jl` v1.5 doesn't use `Tar.jl`, it doesn't explain the weird permissions when unpacking

**Comment by barche** [comment]
I'm not sure if this reproduces the way Pkg does the extraction, but using the sequence below produces a removable tree even with the old version:

```julia
(@v1.5) pkg> add https://github.com/JuliaIO/Tar.jl.git#1b63f2a
    Cloning git-repo `https://github.com/JuliaIO/Tar.jl.git`
  Resolving package versions...
 Installing known registries into `~/tmp/clean_julia_depot_qt`
######################################################################## 100.0%
      Added registry `General` to `~/tmp/clean_julia_depot_qt/registries/General`
  Installed ArgTools ─ v1.1.1
Updating `~/tmp/clean_julia_depot_qt/environments/v1.5/Project.toml`
  [a4e569a6] + Tar v1.8.0 `https://github.com/JuliaIO/Tar.jl.git#1b63f2a`
Updating `~/tmp/clean_julia_depot_qt/environments/v1.5/Manifest.toml`
  [0dad84c5] + ArgTools v1.1.1
  [a4e569a6] + Tar v1.8.0 `https://github.com/JuliaIO/Tar.jl.git#1b63f2a`
  [ea8e919c] + SHA

julia> using Tar
[ Info: Precompiling Tar [a4e569a6-e804-4fa4-b0f3-eef7a1d5b13e]

julia> Tar.extract("/Users/user/Downloads/Qt.v5.15.2.x86_64-apple-darwin.tar", "/Users/user/tmp/tartest")
"/Users/user/tmp/tartest"
```

**Comment by StefanKarpinski** [comment]
The Artifacts code explicitly makes everything readonly:

https://github.com/JuliaLang/Pkg.jl/blob/becccb374edaccb662b106c305a77b1db4a64079/src/Artifacts.jl#L60

@ViralBShah, is your problem that `rm` is complaining that you can't delete those paths because of permissions?

**Comment by StefanKarpinski** [comment]
If so, you had already reported this here https://github.com/JuliaPackaging/Yggdrasil/issues/1969 and it was already discussed that Pkg changes artifacts to be read-only. Is there some other permission issue I'm not seeing here or is that the only one. It's getting a bit hard to tell what this issue is about.

**Comment by ViralBShah** [comment]
Yes, I had reported this in #1969, but my understanding was that the issue was fixed. Hence when it continued to occur with a new release of the `Qt_jll` package, I mentioned it here. I can switch the discussion to #1969. The issue is literally that of permissions of artifacts, and it only happens for `Qt_jll`.

**Comment by StefanKarpinski** [comment]
Issues on Yggdrasil are not helpful for a Pkg problem. But if there's no issue opened over there I guess I don't have to worry about it 🤷🏼‍♂️

**Comment by ViralBShah** [comment]
It is difficult for me to know whether the problem is in the Yggdrasil BinaryBuilder side during creation, or a Pkg problem during the installation, or an issue with Tar. Hence, I'm filing these in Yggdrasil for now, and @giordano has been great with helping figure out what needs to go where.

**Comment by barche** [comment]
> The Artifacts code explicitly makes everything readonly:

OK, after adding the `set_readonly` I can't `rm -rf` the dir anymore, but it seems @giordano is on it in #1969 :)

**Comment by giordano** [comment]
https://github.com/JuliaLang/Pkg.jl/pull/2286

**Comment by ig-or** [comment]
@barche, @giordano,  can I ask why OpenSSL is disabled for QT? is there some specific reason for this?

**Comment by barche** [comment]
Nothing special is done to enable or disable OpenSSL, what functionality do you need, specifically?

**Comment by ig-or** [comment]
@barche, in a file `QtNetwork/qtnetwork-config.h` I noticed 
`#define QT_NO_OPENSSL true`
`#define QT_FEATURE_ssl -1`

from [https://doc.qt.io/qt-5/ssl.html](https://doc.qt.io/qt-5/ssl.html) it's not quite clear for me how this happened..   Maybe adding `OpenSSL_jll` as a dependency and adding  `-openssl-linked`  to `configure` would fix this?

The specific functionality I need is that _I suspect_ marble_jll package is not working as it should because of this, it relies on QT's OpenSSL capability.

**Pull request merged**